crates/proto/src/serialize/txt/trust_anchor.rs

```diff
@@ -0,0 +1,273 @@
+//! Trust anchor file parser
+//!
+//! A trust anchor file largely adheres to the syntax of a zone file but may only contain
+//! DNSKEY records
+
+use std::{borrow::Cow, str::FromStr as _};
+
+use crate::{
+    rr::{dnssec::rdata::DNSKEY, DNSClass, Name, Record, RecordType},
+    serialize::txt::ParseErrorKind,
+};
+
+use super::{
+    rdata_parsers::dnskey,
+    zone,
+    zone_lex::{Lexer, Token as LexToken},
+    ParseError, ParseResult,
+};
+
+/// Trust anchor file parser
+pub struct Parser<'a> {
+    lexer: Lexer<'a>,
+}
+
+impl<'a> Parser<'a> {
```

Comment:
Ugh, all of this duplicate code is unfortunate. I suppose that's my fault for suggesting it should be a separate parser. Do you think there's a way we can abstract all of this stuff over the types of records? Otherwise I might go back on the idea that we should have a separate parser...

Reply:
this trust anchor file pasrer is a lot simpler -- it does not support features like `INCLUDE`, `@`, etc. --  so I actually like that it's a separate, smaller parser. note that I did reuse the `Lexer` and filtered out tokens not allowed in a trust anchor file. the new parser is also relatively well covered by tests

the ways I see to maximize code reuse are:

(1) add a `Mode` field to the existing `Parser` that admits two variants `Zone` and `TrustAnchor`. then we pepper in some branches into the existing `Parser` state machine to change the control flow based on the `Mode` the `Parser` was initialized with

So that may look like going from

``` rust
self.state = match self.state {
    State::StartLine => {
        match token {
            Token::A | token::B => return Err(UnsupportedToken);
            Token::C => State::X,
            Token::D => State::Y,
            // ..
        }
    }
    // ..
};
```

to


``` rust
self.state = match self.state {
    State::StartLine => {
        match token {
            Token::A | token::B => return Err(UnsupportedToken);
            Token::C if mode.is_zone() => State::X,
            Token::C => return Err(UnsupportedToken),
            Token::D if mode.is_zone() => State::Y,
            Token::D => State::Z,
            // ..
        }
    }
    // ..
};
```

which makes the state machine harder to read and harder to modify. I would not do that kind of changes without first adding a lot more unit tests that serve as regression tests for the zone file parser.

(2) the other way is to have a `TrustAnchorParser` that it's a newtype over `Parser` whose `parse` method calls into `Parser::parse` and then filters out any record that is not allowed by the trust anchor file parser -- it effectively behaves like a filter from `Map<RecordKey, Record>` to `Vec<Record<DNSKEY>>`.

... it's what I initially thought but I don't think it would be that straightforward because we don't want `TrustAnchorParser` to support `INCLUDE` nor `@` and if you have its `parse` method call into `Parser::parse` then those features would have been sort of already "expanded" and `TrustAnchorParser` would not be able to filter out or undo the effect of those features.

So in this route, it would also be necessary to add some branching to `Parser` to reject those features in `Parser::parse`

---

but overall I think (1) and (2)'s downsides outweigh the code reuse gain

Reply:
Fair enough.